components/style/values/specified/calc.rs

```diff
@@ -0,0 +1,549 @@
+                    }
+                }
+                _ => {
+                    input.reset(position);
+                    break
+                }
+            }
+        }
+
+        Ok(root)
+    }
+
+    /// Parse a top-level `calc` expression, and all the products that may
+    /// follow, and stop as soon as a non-product expression is found.
+    ///
+    /// This should parse correctly:
+    ///
+    ///     * `2`
+    ///     * `2 * 2`
+    ///     * `2 * 2 + 2` (but will leave the `+ 2` unparsed).
+    ///
+    fn parse_product(
+        context: &ParserContext,
+        input: &mut Parser,
+        expected_unit: CalcUnit)
+        -> Result<Self, ()>
+    {
+        let mut root = Self::parse_one(context, input, expected_unit)?;
+
+        loop {
+            let position = input.position();
+            match input.next() {
+                Ok(Token::Delim('*')) => {
+                    let rhs = Self::parse_one(context, input, expected_unit)?;
+                    let new_root = CalcNode::Mul(Box::new(root), Box::new(rhs));
+                    root = new_root;
+                }
+                // TODO(emilio): Figure out why the `Integer` check.
+                Ok(Token::Delim('/')) if expected_unit != CalcUnit::Integer => {
+                    let rhs = Self::parse_one(context, input, expected_unit)?;
+                    let new_root = CalcNode::Div(Box::new(root), Box::new(rhs));
+                    root = new_root;
+                }
+                _ => {
+                    input.reset(position);
+                    break
+                }
+            }
+        }
+
+        Ok(root)
+    }
+
+    /// Tries to simplify this expression into a `<length>` or `<percentage`>
+    /// value.
+    fn to_length_or_percentage(&self) -> Result<CalcLengthOrPercentage, ()> {
+        let mut ret = CalcLengthOrPercentage::default();
+        self.add_length_or_percentage_to(&mut ret, 1.0)?;
+        Ok(ret)
+    }
+
+    /// Puts this `<length>` or `<percentage>` into `ret`, or error.
+    ///
+    /// `factor` is the sign or multiplicative factor to account for the sign
+    /// (this allows adding and substracting into the return value).
+    fn add_length_or_percentage_to(
+        &self,
+        ret: &mut CalcLengthOrPercentage,
+        factor: CSSFloat)
+        -> Result<(), ()>
+    {
+        match *self {
+            CalcNode::Percentage(pct) => {
+                ret.percentage = Some(ret.percentage.unwrap_or(0.) + pct * factor)
+            }
+            CalcNode::Length(ref l) => {
+                match *l {
+                    NoCalcLength::Absolute(abs) => {
+                        // FIXME(emilio): Can we do better the factor dance
+                        // here without complicating the code too much?
+                        ret.absolute = Some(
+                            ret.absolute.unwrap_or(Au(0)) +
+                            Au::from_f32_px(Au::from(abs).to_f32_px() * factor)
```

Comment:
It also looks like `NoCalcLength` impls `Mul<CSSFloat>` for scaling stuff. Not sure how useful would that be, since we still need to match the variants.